ReSharper.FSharp/src/FSharp/FSharp.Psi.Features/src/FSharpTypingAssist.fs

```diff
@@ -743,42 +742,53 @@ type FSharpTypingAssist(lifetime, dependencies) as this =
                 lexer.Advance(-1)
             lexer.TokenEnd
 
-        match x.GetFSharpTree(textControl) with
-        | None -> false
-        | Some parseTree ->
-
         let document = textControl.Document
-        let visitor =
-            { new SyntaxVisitorBase<_>() with
-                member x.VisitExpr(_, _, defaultTraverse, expr) =
-                    match expr with
-
-                    // if expr then {caret}
-                    // No info is available after error recovery, we check that range is surrounded with if ... then.
-                    | SynExpr.FromParseError (ExprRange range, _)
 
-                    // while expr do {caret}
-                    | SynExpr.While (_, ExprRange range, _, _) when
-                            getEndOffset document range = prevTokenEnd ->
-                        Some range
+        let fsFile = textControl.GetFSharpFile(dependencies.Solution)
+        let expr = fsFile.GetNode<IFSharpExpression>(DocumentOffset(document, lexer.TokenStart))
+        if isNull expr then false else
 
-                    // for i = ... to ... do {caret}
-                    | SynExpr.For (_, _, IdentRange range, _, _, _, _, ExprRange lastRange, _)
-
-                    // for ... in ... do {caret}
-                    | SynExpr.ForEach (_, _, _, _, PatRange range, ExprRange lastRange, _, _) when
-                            getEndOffset document lastRange = prevTokenEnd ->
-                        Some range
-
-                    | _ -> defaultTraverse expr }
+        let range =
+            let inline checkExpr (expr: IFSharpExpression) (node: ITreeNode) =
```

Comment:
While I was coming up with a name for the function, I found an old bug for the case
```fsharp
for a = 0
        to 10 do  {caret}
```
its' because here
![image](https://github.com/user-attachments/assets/f330aa73-91e3-494d-9824-6448c10cef35)
the range was taken not from `to ... ' expr, but `do ...`, which could never be matched with prevTokenEnd